lib/Models/RegionProvider.js

```diff
@@ -0,0 +1,405 @@
+        // ##TODO think about whether this matters for disambigs
+        numericCodes = true;
+        codes = dataset.getVariableValues(regionVariable);
+    }
+    var results = {
+        regionValues: new Array(this.regions.length),
+        ambiguousMatches: {},
+        failedMatches: {},
+        successes: 0,
+        totalRows: 0
+    };
+
+    var indices = (rowList ? rowList : Object.keys(codes)); // iterate either over every row, or just these rows
+    var regionVar = dataset.variables[dataset.getRegionVariable()];
+
+    for (var i = 0; i < indices.length; i++) {
+        var row = indices[i];
+        results.totalRows++;
+        var index = this.findRegionIndex(codes[row], disambigCodes[row]); // will pass ,undefined if no disambig.
+
+        if (index < 0) {
+            results.failedMatches[codes[row]] = true; //##TODO should combine disambig into the key?
+            continue;
+        }
+        // not really "ambiguous" - actually if two data values for one region.
+        if (results.regionValues[index] !== undefined && !dataset.hasTimeData()) {
+            // if there is a time column, ignore ambiguous matches
+            results.ambiguousMatches[codes[row]] = true;
+            continue;
+        }
+        if (defined(vals)) {
+            results.regionValues[index] = vals[row];
+        } else {
+            results.regionValues[index] = 0; // In some cases region mapping without a data variable is meaningful.
+        }
+        if (defined(regionVar)) {
+            // set reverse lookup from variable back to region index.
+            regionVar.setRegionCode(i, this.regions[index].id);
+        }
+        results.successes ++;
+    }
+    return results;
+};
+
+/**
+ * Pre-generates a function which quickly turns a value into a colour.
+ * @param {Number[]} regionValues Array of values.
+ * @param {Function} colorFunc should be function(val) { return [r,g,b,a]; } 
+ * @returns {Function} Function of type f(regionIndex) { return [r,g,b,a]; }
+ */
+RegionProvider.prototype.getColorLookupFunc = function(regionValues, colorFunc) {
+    var colors = regionValues.map(colorFunc);
+    return function(regionIndex) {
+        // actually maybe it's ok to return undefined after all
+        return colors[regionIndex] !== undefined ? colors[regionIndex] : [0,0,0,0]; //# is it important to get rid of this if?
+    };
+
+};
+
+
+
+/*
+ * @returns The index of the first column that left-matches any of the given aliases 
+ * @param varNAmes Array of variable names.
+ */
+RegionProvider.prototype.findRegionVariable = function(varNames) {
+    for (var j = 0; j < this.aliases.length; j++) {
+        for (var i = 0; i < varNames.length; i++) {
+            //check that it is the right type of code
+            //if (typeof this.regionTextCode === "boolean" && this.regionTextCode !== variables[i].isTextCode) {
+            //    continue;
+            //}
+            // SB: I don't think that's a worthwhile test
```

Comment:
I don't understand the question well enough to say if I agree with you, but if you're sure then just delete the code and comment.